### ISSUE TITLE
New cop to detect unnecessary excessive guards

### ISSUE THREAD

**Issue by fatkodima** [rubocop/rubocop#9018]
Its sure debatable thing, but I personally find such code
https://github.com/rubocop-hq/rubocop/blob/d13acfb1910b58907fcf77d412b05f2f1801ed33/lib/rubocop/cop/message_annotator.rb#L114-L121
harder to parse and understand. For every guard you should shift to the modifier conditional and then return back to the [redundant, if rewritten] `return`, breaking the nice top-down navigation through the code, rewritten as

```ruby
def display_cop_names?
  if debug?
    true
  elsif options[:display_cop_names] == false
    false
  elsif options[:display_cop_names]
    true
  elsif options[:format] == 'json'
    false
  else
    config.for_all_cops['DisplayCopNames']
  end
end
```

Another example
https://github.com/rubocop-hq/rubocop/blob/d13acfb1910b58907fcf77d412b05f2f1801ed33/lib/rubocop/cop/autocorrect_logic.rb#L28-L37

which can be rewritten as
```ruby
def autocorrect_enabled?
  if !cop_config
    # allow turning off autocorrect on a cop by cop basis
    true
  elsif cop_config['AutoCorrect'] == false
    false
  elsif @options.fetch(:safe_auto_correct, false)
    safe_autocorrect?
  else
    true
  end
end
```

@bbatsov @marcandre Wdyt on this style?

**Comment by koic** [comment]
I'm worried about the possibility of false changing the code style intended by writer when an intended (many) pre-guard conditions and an essential processing are separated.

**Comment by marcandre** [comment]
Interesting question. When I first started to work in the RuboCop codebase, I thought all these `return` were very surprising and didn't like them. My typical style was to avoid using `return` completely.

Since then, I've surprised myself in liking it 😅 . It switched in my mind to 'hey, I just figured out unequivocally what I need to answer, let's just get out of here right now'.

I'd say my ability to read these improved a lot. I still can't read `elsif` and avoid them.

**Comment by bbatsov** [comment]
> Since then, I've surprised myself in liking it 😅 . It switched in my mind to 'hey, I just figured out unequivocally what I need to answer, let's just get out of here right now'.

Same here. Probably none of you knows the backstory, but this came from one of the first cops that was originally designed to target the problem of deeply nested `if`s that were the only/final expression in a method definition. Admittedly using `return` to flatten the conditionals wasn't popular in Ruby back then, but to my surprise a lot of people seemed to like it. 

That being said, I realize that there are people who share @fatkodima's perspective, so I'd be fine to add such a cop, but it should probably be disabled by default.
